### PR TITLE
Download clicky data

### PR DESCRIPTION
**Pull request by sarahmcdougall**
# Summary
Creates a GitHub workflow that runs a Python script each Sunday and the first of each month. The script gathers the latest visitor data from Clicky and adds it to the existing visitor data in the repository. The script then creates a JSON file of the updated visitor data (`clickyData.json`) and a JPEG of a Matplotlib representation of the data (`clickyGraph.jpg`).

# Testing Guidance
* If you have not done so already, follow the [MITRE Python SSL Setup](https://awesome.pages.mitre.org/full-stack-mitre/dev-environment/programming-languages/python/) to install the MITRE certs into your Python environment.
* Make sure you have the necessary imports for running the script - `requests` and `matplotlib` (you may already have them installed)
* Run the `runClickyScript.py` file within the Python shell and check that the script runs successfully. Then, check that the files `clickyData.json` and `clickyGraph.jpg` are created. The `__pycache__` directory will also get created, but it should be git ignored.

### CODE REVIEW

File: clicky/clickyScript.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        
+'''
+Retrieves the new visitor data from the Clicky url.
+
+Returns: python list containing new data from the clicky url
+'''
+def getNewData(clickyUrl):
+    # this is a bandaid solution for the SSL, requires further attention
+    # would need the certificates to be in place and removal of "verify=False"
+    response = requests.get(clickyUrl)
```
Comment (mickohanlon23) [inline clicky/clickyScript.py:37]: It might be helpful to add a try/catch here to throw a user-friendly exception if there is an error connecting to the Clicky API.

File: clicky/clickyScript.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+'''
+Appends the relevant parts of the new visitor data to the existing
+visitor data currently checked into the repository. This includes
+the new number of visits, new IP addresses, new unique visits, and
+the dates of the new visits.
+
+Returns: the merged data
+'''
+def merge(newData, oldData):
+    numUniqueVisits = []
```
Comment (mickohanlon23) [inline clicky/clickyScript.py:51]: We might want to store the unique IP address somewhere in the stored `clickyData.json`. I don't think we necessarily care how many unique visitors visit within each week's span, but rather how many new visitors visit each week.
Otherwise put, comparing each `newData` entry to the entire list of `ipAddresses` that have ever visited the site versus just comparing it to other ipAddresses within the new week's data.

Note, this would also require changing the label in the `plotData()` function to something like "New Visits"

File: clicky/runClickyScript.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from clickyScript import *
+
+
+clickyUrl = "https://api.clicky.com/api/stats/4?site_id=101369228&sitekey=6d6a506f44d45a59&type=visitors-list&date=last-week&output=json"
```
Comment (mickohanlon23) [inline clicky/runClickyScript.py:4]: I wonder if we should move the `site_id` and `sitekey` parameters to environment variables? We already have the `site_id` in the `.env` file.

- Our site details: https://clicky.com/stats/prefs?site_id=101369228
- Info on `site_id and `sitekey`: https://clicky.com/help/api

It looks like like `sitekey` is almost intended as a password to access the API.

File: clicky/clickyScript.py (hunk 1 above)
Comment (mickohanlon23) [reply clicky/clickyScript.py:37]: We can probably also delete the comment above if we're no longer adding the flag to disable SSL.

File: .github/workflows/clicky-workflow.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        id: check
+        run: |
+          if git diff --exit-code; then
+            echo "::set-output name=diff::false"
+          else
+            echo "::set-output name=diff::true"
+          fi
+      - 
+        name: "Push changes"
+        if: steps.check.outputs.diff == 'true'
```
Comment (mickohanlon23) [inline .github/workflows/clicky-workflow.yml:41]: Something seems to be not quite working in the logic of this conditional. This step of the workflow wasn't previously running, but removing this line allowed for the commit to successfully be made.

File: clicky/clickyScript.py (hunk 1 above)
Comment (sarahmcdougall) [reply clicky/clickyScript.py:37]: Good point. I started looking into the request exceptions a little more in the [requests documentation](https://requests.readthedocs.io/en/latest/user/quickstart/#errors-and-exceptions). Based on this [stack overflow thread](https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module), it seems common to either have custom handling for the different types of exceptions, or you can catch the base-class exception for requests. Do you think it would be most appropriate to catch the base-class exception and then raise a `SystemExit` error? Or should we implement a re-try loop if there is a connection error, etc.? Not sure if we need to be that thorough with the exception handling or if catching the base class exception would be good enough here

File: .github/workflows/clicky-workflow.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      - 
+        name: "Install dependencies"
+        run: |
+            python -m pip install --upgrade pip
+            pip install requests matplotlib
+      - 
+        name: "Run clicky script"
+        working-directory: "clicky"
+        run: |
+            python clickyScript.py
```
[the remark this answers is not in the document]
Comment (mickohanlon23) [reply .github/workflows/clicky-workflow.yml]: Fixed in https://github.com/dvci/shc-web-verifier/pull/81/commits/24c6c6448491ae1fa39c5ba97dc39273e54b641f

File: clicky/clickyScript.py (hunk 1 above)
Comment (mickohanlon23) [reply clicky/clickyScript.py]: I agree with the way you implemented it by just checking the base class. I don't think there is a need to specifically handle each type of request error - especially since these errors will only show up in the GitHub Action so most likely we will be the only ones actually seeing them.

Comment (mickohanlon23) [reply clicky/clickyScript.py]: We may want to include lines 43 and 44 in the try block. The Clicky site doesn't necessarily throw an error status if the data is unable to be retrieved. 

For example, the Clicky site has 404 error handling so that a GET request to something like https://clicky.com/not-a-page actually returns a 200 status but renders a "404 page not found" human-readable page, triggering an error in the `json = response.json()` call. Alternatively, modifying the site key (for example https://api.clicky.com/api/stats/4?site_id=1019228&sitekey=6d6af44d45a59&type=visitors-list&date=last-week&output=json) will in fact return a JSON object, but it will not have the expected fields, leading to an error in the `newDataList = json[0]['dates'][0]['items']` call.

Long story short, I think it probably makes sense for any response that doesn't include a JSON body of the right schema to be captured in a common exception.

File: clicky/runClickyScript.py (hunk 3 above)
Comment (mickohanlon23) [reply clicky/runClickyScript.py]: Awesome! I just tested locally and via the GitHub Action on my cloned repo, and both scenarios were still able to access the correct URL using the environment variables.

File: .github/workflows/clicky-workflow.yml (hunk 4 above)
Comment (mickohanlon23) [reply .github/workflows/clicky-workflow.yml]: I saw you removed this check, along with the check dirty files step.

File: clicky/clickyScript.py [hunk 6]
```diff
@@ -70,8 +76,8 @@ def plotData(allData):
     plt.plot(X1, Y1, 'o-', label = 'Total Visits')
```
Comment (mickohanlon23) [inline clicky/clickyScript.py:76]: It just occurred to me that "Total Visits" could be a bit misleading / vague. I think it would help to make this data label 'Total Weekly Visits'. Adding "weekly" to the Y-axis label and plot title might help as well.

File: clicky/clickyScript.py (hunk 2 above)
Comment (mickohanlon23) [reply clicky/clickyScript.py]: I like the new logic. I tested out manually generating a graph by creating some fake data in a local `clickyData.json` file. The new graph is generated as expected and clearly portrays the data we want to gather. Thanks for updating!

Pasting graph here in case other want to review:
clickyGraph

**Review by mickohanlon23 — APPROVED** [review]
New updates successfully generate the graph and data files in the default branch!

The primary original issues have been addressed. There are two outstanding comments suggesting very minor changes, but other than that I approve. Thanks for wrapping this up!

File: clicky/clickyScript.py (hunk 6 above)
Comment (mickohanlon23) [reply clicky/clickyScript.py]: Thanks!
